Add aws_acm module

matt-telstra opened the pull request:
##### SUMMARY

This satisfies the feature request in #60120 

There currently exists the `aws_acm_facts` module. (Which appears to have been renamed to `aws_acm_info` in the devel branch.) This fetches data about certificates in Amazon's certificate manager (ACM).

This PR adds a new module, `aws_acm`, which is used to create, modify and delete ACM resources. (Only self-provided certificates for now, not the Amazon-managed ones.)

As part of this I have:

* factored out the bulk of `aws_acm_info` into a `module_util`, so it can be used by both `aws_acm` and `aws_acm_info`.
* converted `aws_acm_info` to an AWS Ansible module. Previously it was a generic Ansible module.
* Added integration tests for `aws_acm_info` and `aws_acm`. (It makes sense to test them together. You can't test cert creation unless you can get cert info. And you can't test getting cert info unless you create a cert.)
* added more filtering options for `aws_acm_info`

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

aws_acm
aws_acm_info

##### ADDITIONAL INFORMATION

I have written a unit test for `aws_acm`. When I manually run it with `ansible-test units --docker --python 2.7` it works. Can someone please double check that the CICD will invoke this test? (I just added the file. I don't know whether I need to point something towards it.)

**The integration test currently does not pass.**
I need help debugging why.

When I ran it manually as a standalone playbook it did pass.
But now I get:


`ansible-test integration --docker ubuntu1604 aws_acm`

```
TASK [aws_acm : Find chains] ***************************************************
 [WARNING]: Failure using method (v2_runner_item_on_failed) in callback plugin
(<ansible.plugins.callback.default.CallbackModule object at 0x7f369c13d510>):
'NoneType' object has no attribute 'strip'
```

What does that mean? This task succeeds when run as a normal playbook. So I have the right syntax.
This task uses `certificate_complete_chain`. Was a bug introduced to `certificate_complete_chain` in the devel branch? When I run the integration test for `certificate_complete_chain`, every meaningful task in the playbook is skipped by Ansible.

I am also not certain that I have added all required permissions to `hacking/aws_config/testing_policies/security-policy.json`

Tagging @willthames 

(Note that @mlda065 and @matt-telstra are the same person. I'm switching account for the pull request because this is a contribution on behalf of my employer.)

matt-telstra committed:
convert aws_acm_facts to AnsibleAWSModule

matt-telstra committed:
factor aws_acm_facts into module_utils

matt-telstra committed:
add more filtering options for aws_acm_info

matt-telstra committed:
add aws_acm module and tests

matt-telstra commented:
Also, how can I have a look at the html documentation for this module? I want to check that the yaml strings inside python variables are converted to html as expected.

matt-telstra committed:
uncomment aws_acm test

willthames commented:
```
pip install -r docs/docsite/requirements.txt
MODULES=aws_acm make webdocs
```
will create the relevant page in docs/docsite/_build/html/modules

matt-telstra commented:
I added some fields to `aws_acm_info`. So should I increment the `metadata_version`?

matt-telstra commented:
bot_broken

The "[explain](https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/yamllint.html)" link takes me to a page which contains no explanation of anything. Just a title and no body.

willthames commented:
@matt-telstra metadata_version should [be 1.1](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#ansible-metadata-block)

willthames commented:
I'm not sure what's going on with `version_added` checks right now (I wonder if someone prematurely bumped it to 2.10 and it's now back to 2.9 - 2.9 seems good though)

willthames commented:
If there are any other test failures that you don't understand, let me know - most of them are just standard PEP8, the documentation test failures are relatively well explained but some are less obvious than others.

matt-telstra committed:
fix linting for aws_acm

matt-telstra commented:
I deliberately set `version_added` to the next release, since I assume this module won't be merged and published until the next release. 2.9 is already out, right?

willthames commented:
@matt-telstra it's not impossible for it to get into 2.9 - there are a couple of weeks before beta feature freeze, as I understand it

matt-telstra committed:
fix __future__ linting for aws_acm

matt-telstra committed:
fix linting for aws_acm

matt-telstra committed:
fix linting for aws_acm

matt-telstra committed:
fix linting for aws_acm

matt-telstra committed:
fix linting for aws_acm

matt-telstra commented:
I'm getting this error:

```
00:45 ERROR: Found 1 future-import-boilerplate issue(s) which need to be resolved:
00:45 ERROR: test/sanity/ignore.txt:534:1: Ignoring 'lib/ansible/modules/cloud/amazon/aws_acm_info.py' is unnecessary (75%)
```

I have not touched `ignore.txt`. I'm assuming this was put in because `aws_acm_info` has no tests.

Is it safe to delete the following lines from `ignore.txt`?

```
lib/ansible/modules/cloud/amazon/aws_acm_info.py future-import-boilerplate
lib/ansible/modules/cloud/amazon/aws_acm_info.py metaclass-boilerplate
lib/ansible/modules/cloud/amazon/aws_acm_info.py validate-modules:E337
lib/ansible/modules/cloud/amazon/aws_acm_info.py validate-modules:E338
```

Note that I'm getting errors about the first two lines, not the second two.

willthames commented:
@matt-telstra you can delete the second set of two lines but then you have to deal with the results:

```
18:15 ERROR: Found 3 validate-modules issue(s) which need to be resolved:
18:15 ERROR: lib/ansible/modules/cloud/amazon/aws_acm_info.py:0:0: E337 Argument 'statuses' in argument_spec defines type as 'list' but documentation doesn't define type (75%)
18:15 ERROR: lib/ansible/modules/cloud/amazon/aws_acm_info.py:0:0: E338 Argument 'certificate_arn' in argument_spec uses default type ('str') but documentation doesn't define type (75%)
18:15 ERROR: lib/ansible/modules/cloud/amazon/aws_acm_info.py:0:0: E338 Argument 'domain_name' in argument_spec uses default type ('str') but documentation doesn't define type (75%)
```
which just means adding `type` to the documentation of the parameters

matt-telstra committed:
fix aws_acm_info arg type

matt-telstra commented:
Ah, I see.

(I only just noticed the "test" tab in Shippable. Previously I was looking at the "Console" tab, which is why I took so many commits to fix those linting issues.)

willthames commented on lib/ansible/module_utils/aws/acm.py:
```diff
@@ -0,0 +1,210 @@
+        self.module = module
+
+        region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module, boto3=True)
+        self.client = boto3_conn(module, conn_type='client',
```
just do module.client('acm') these days

matt-telstra replied:
Fixed

willthames commented on lib/ansible/module_utils/aws/acm.py:
```diff
@@ -0,0 +1,210 @@
+        module.debug("Attempting to delete certificate %s" % arn)
+        try:
+            self.delete_certificate_with_backoff(client, arn)
+        except botocore.exceptions.ClientError as e:
```
need to deal with `BotoCoreError` in all exception handling

matt-telstra replied:
What do you mean? That code is how the existing boto calls were made. That `except` statement catches a botocore error.

Or do you mean I need to change both the new and the old code from `except botocore.exceptions.ClientError` to `except botocore.exceptions.BotoCoreError`?

matt-telstra replied:
Ah, I looked at another module and saw what you mean. It's fixed now.

willthames commented on test/integration/targets/aws_acm/aliases:
```diff
@@ -0,0 +1,4 @@
+cloud/aws
+shippable/aws/group2
+aws_acm_info
+aws_acm_facts
```
no need to test `aws_acm_facts` (if there is it should be `_aws_acm_facts`)

matt-telstra replied:
Fixed

matt-telstra committed:
remove test for old module name aws_acm_facts

matt-telstra committed:
simplify AWS ACM client creation

willthames commented on test/integration/targets/aws_acm/tasks/main.yml:
```diff
@@ -0,0 +1,457 @@
+  block:
+
+  - set_fact:
+    aws_connection_info: &aws_connection_info
```
Needs indenting so that aws_connection_info is  a parameter to set_fact

matt-telstra replied:
Ah, that's embarrassing. (It's due to a recent move of the `set_fact` task into the block)

matt-telstra replied:
I've fixed it. (I'll push the commit in a few minutes)

willthames commented on test/integration/targets/aws_acm/meta/main.yml:
```diff
@@ -0,0 +1,3 @@
+dependencies:
+  - setup_openssl
```
This doesn't seem to work very well in the ansible-test container:

```
TASK [setup_openssl : Install pyOpenSSL] ***************************************
task path: /root/.ansible/test/tmp/aws_acm-t_0b90zn-ÅÑŚÌβŁÈ/test/integration/targets/setup_openssl/tasks/main.yml:6
<testhost> ESTABLISH LOCAL CONNECTION FOR USER: root
<testhost> EXEC /bin/sh -c 'echo ~root && sleep 0'
<testhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1565867400.6824777-102379632105809 `" && echo ansible-tmp-1565867400.6824777-102379632105809="` echo /root/.ansible/tmp/ansible-tmp-1565867400.6824777-102379632105809 `" ) && sleep 0'
Using module file /root/ansible/lib/ansible/modules/packaging/os/apt.py
<testhost> PUT /root/.ansible/tmp/ansible-local-132q9lvaufd/tmpdh2r6nem TO /root/.ansible/tmp/ansible-tmp-1565867400.6824777-102379632105809/AnsiballZ_apt.py
<testhost> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1565867400.6824777-102379632105809/ /root/.ansible/tmp/ansible-tmp-1565867400.6824777-102379632105809/AnsiballZ_apt.py && sleep 0'
<testhost> EXEC /bin/sh -c 'ANSIBLE_DEBUG_BOTOCORE_LOGS=True /tmp/python-qme69j5a-ansible/python /root/.ansible/tmp/ansible-tmp-1565867400.6824777-102379632105809/AnsiballZ_apt.py && sleep 0'
<testhost> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1565867400.6824777-102379632105809/ > /dev/null 2>&1 && sleep 0'
 [WARNING]: Updating cache and auto-installing missing dependency: python3-apt

The full traceback is:
  File "/tmp/ansible_apt_payload_n70gyc60/__main__.py", line 1038, in main
    import apt
  File "/usr/lib/python3/dist-packages/apt/__init__.py", line 23, in <module>
    import apt_pkg

fatal: [testhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "allow_unauthenticated": false,
            "autoclean": false,
            "autoremove": false,
            "cache_valid_time": 0,
            "deb": null,
            "default_release": null,
            "dpkg_options": "force-confdef,force-confold",
            "force": false,
            "force_apt_get": false,
            "install_recommends": null,
            "name": "python3-openssl",
            "only_upgrade": false,
            "package": [
                "python3-openssl"
            ],
            "policy_rc_d": null,
            "purge": false,
            "state": "present",
            "update_cache": null,
            "upgrade": null
        }
    },
    "msg": "Could not import python modules: apt, apt_pkg. Please install python3-apt package."
}

PLAY RECAP *********************************************************************
testhost                   : ok=2    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

honestly never used this role

The other problem with it is I later get another warning:
```
ERROR: Test target "aws_acm" has changed the test environment!
If these changes are necessary, they must be reverted before the test finishes.
```
This should probably all be done in a virtual environment (perhaps rather than having setup_openssl as a dependency you could include the role after setting up a virtualenv in tasks/main.yml)

matt-telstra replied:
Where did you see that error message?

Why does it not work? This is exactly how the crypto modules I am using do their tests.

How do I set up a virtual environment within a playbook, and run the subsequent tasks _within_that environment?

willthames replied:
I'm running `ansible-test integration --docker -vvv aws_acm` and that's what happens.
 
See e.g. ecs_cluster's test suite for how to run a test suite inside a virtualenv

matt-telstra replied:
Ok, I just pushed a new commit to do that. When I run that command I get an authorization error.

> Couldn't obtain certificates: An error occurred (UnrecognizedClientException) when calling  the ListCertificates operation: The security token included in the request is invalid", resource_actions": ["acm:ListCertificates"].

I put some AWS creds in `cloud-config-aws.ini`. Was there something else I have to do?

willthames replied:
Just ensure security_token is set to "" if one isn't required.

matt-telstra replied:
I already did that.

matt-telstra replied:
```
# This is the configuration template for ansible-test AWS integration tests.
#
# You do not need this template if you are:
#
# 1) Running integration tests without using ansible-test.
# 2) Using the automatically provisioned AWS credentials in ansible-test.
#
# If you do not want to use the automatically provisioned temporary AWS credentials,
# fill in the @VAR placeholders below and save this file without the .template extension.
# This will cause ansible-test to use the given configuration instead of temporary credentials.
#
# NOTE: Automatic provisioning of AWS credentials requires an ansible-core-ci API key.

[default]
aws_access_key: ABCDE
aws_secret_key: 123456
security_token: ""
aws_region: ap-southeast-2
# aws_cleanup controls whether the environment is cleaned up after tests have completed
# This only applies to tests that have a cleanup stage
# Defaults to true when using this template
# aws_cleanup: true
# aliases for backwards compatibility with older integration test playbooks
ec2_access_key: {{ aws_access_key }}
ec2_secret_key: {{ aws_secret_key }}
ec2_region: {{ aws_region }}
```

willthames replied:
Sorry, I wasn't on my test machine. Apparently it's `security_token:` - ie. nothing after the colon

matt-telstra replied:
Yes that `security_token` bit fixed it.

matt-telstra committed:
fix indent typo in aws_acm test

matt-telstra committed:
catch BotoCoreError in aws_acm

matt-telstra committed:
fix indent typo in aws_acm test

matt-telstra commented:
I'm getting this error for one of the shippable tests:

```
                          Traceback (most recent call last):
  File "/tmp/ansible_aws_acm_info_payload_WZmZzV/ansible_aws_acm_info_payload.zip/ansible/module_utils/aws/acm.py", line 95, in get_certificates
    all_certificates = self.list_certificates_with_backoff(client=client, statuses=statuses)
  File "/tmp/ansible_aws_acm_info_payload_WZmZzV/ansible_aws_acm_info_payload.zip/ansible/module_utils/cloud.py", line 150, in retry_func
    raise e
ClientError: An error occurred (AccessDeniedException) when calling the ListCertificates operation: User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible=ansible=137540.108 is not authorized to perform: acm:ListCertificates
```

But I already added `acm:ListCertificates` to `hacking/aws_config/testing_policies/security-policy.json`. Do I need to do something else? Or do I need a human to update the IAM role to match this updated policy document?

willthames commented:
@matt-telstra the security-policy.json changes help manually test PRs but also needs a change to https://github.com/mattclay/aws-terminator to update the CI policies.

matt-telstra commented:
Do you mean I need to make a pull request [here](https://github.com/mattclay/aws-terminator/blob/master/aws/policy/security-services.yaml)?

matt-telstra commented:
Because [that file](https://github.com/mattclay/aws-terminator/blob/master/aws/policy/security-services.yaml) looks different to [this file](https://github.com/ansible/ansible/pull/60552/files#diff-29f2eef49ed9909aa266c79db1f5de12)

willthames commented:
@matt-telstra sure, the files are different. But that's where the permissions need to be for CI to work (not all AWS modules have test suites that run on the CI)

matt-telstra committed:
tighten AWS ACM test policy resource

matt-telstra committed:
move aws acm int test to venv

willthames commented on =0.15:
```diff
@@ -0,0 +1,6 @@
+Requirement already satisfied: pyopenssl in ./venv/lib/python3.6/site-packages (19.0.0)
```
I'm not sure this file should be here!

matt-telstra replied:
Ah, damn. I thought I deleted that.

Anyway, it's deleted now.

willthames reviewed (changes requested):
Apart from the errant `=0.15` file, this all looks good to me.

All tests pass locally, so just a matter of waiting for CI to catch up.

matt-telstra committed:
remove errant file

matt-telstra committed:
fix AWS ACM int test perms

felixfontein commented on lib/ansible/modules/cloud/amazon/aws_acm.py:
```diff
@@ -0,0 +1,353 @@
+    while '--' in pem:
+        pem = pem.replace('--', '-')
+
+    return(pem.lower().strip())
```
Don't lower-case PEM. It is Base64 encoded, and lower-casing will change the content!

jillr replied:
There should be helpers in lib/ansible/module_utils/crypto.py that can handle this.

matt-telstra replied:
Ah thanks. I've removed the `.lower()`.

The reason I didn't use a helper is because I don't want to add `openssl` or whatever else as a dependency.

If you believe that the value of using a proper helper outweighs the cost of adding a dependency, then let me know and I'll change it.

matt-telstra replied:
The existing function in that handler requires a private key, which we cannot get for a cert we're pulling from ACM. So I'm writing my own function (which just uses the cryptography library.

Where should I put unit tests for this new function? It doesn't seem right to put them in the ACM unit tests. I can't see any crypto stuff in `test/units/module_utils`.

matt-telstra replied:
I have just added unit tests to the `aws_acm` module unit tests for now. Let me know if I should move them.

felixfontein commented on test/integration/targets/certificate_complete_chain/tasks/main.yml:
```diff
@@ -1,4 +1,21 @@
----
+# test code for the changed_when parameter
```
Is there a reason why you modify `test/integration/targets/certificate_complete_chain/tasks/main.yml` in this way? If I understand this correctly, you are adding your copyright to a file you have not been involved in at all.

matt-telstra replied:
Ah, no. I think I modified the wrong file there. I intended to add the copywrite thing to some other file called `main.py`. I'll undo that.

matt-telstra replied:
I meant to add it to `test/integration/targets/aws_acm/tasks/main.py`. But do I even need it at all there? I see *some* `main.py` integration tests have one, some don't. Why is that?

Now that I think about it, is it even valid to put a copyright notice in any of those files? Isn't the whole repo copyrighted by the project owner? Or is it that I *own* the copyright for the files I contribute, and *license* them under GPLv3?

felixfontein replied:
I think it is the latter: you own the copyright, and license them as GPLv3.

jillr commented on lib/ansible/module_utils/aws/acm.py:
```diff
@@ -0,0 +1,209 @@
+    # A cert with that ARN must already exist
+    def get_domain_of_cert(self, client, module, arn):
+        if arn is None:
+            module.fail(msg="Internal error with ACM domain fetching" % arn)
```
```suggestion
            module.fail(msg="Internal error with ACM domain fetching, no certificate ARN specified")
```

jillr commented on lib/ansible/modules/cloud/amazon/aws_acm_info.py:
```diff
@@ -12,17 +15,34 @@
 description:
   - Retrieve information for ACM certificates
   - This module was called C(aws_acm_facts) before Ansible 2.9. The usage did not change.
+  - Note that this will not return facts about uploaded keys of size 4096 bits, due to a limitation of the ACM API.
```
```suggestion
  - Note that this will not return information about uploaded keys of size 4096 bits, due to a limitation of the ACM API.
```

jillr commented on lib/ansible/modules/cloud/amazon/aws_acm_info.py:
```diff
@@ -226,106 +259,37 @@
     )
-    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
+    module = AnsibleAWSModule(argument_spec=argument_spec, supports_check_mode=True)
+    acm_facts = ACMServiceManager(module)
```
```suggestion
    acm_info = ACMServiceManager(module)
```

jillr commented on lib/ansible/modules/cloud/amazon/aws_acm_info.py:
```diff
@@ -226,106 +259,37 @@
                         region=region, endpoint=ec2_url, **aws_connect_kwargs)
 
-    certificates = get_certificates(client, module, domain_name=module.params['domain_name'], statuses=module.params['statuses'])
+    certificates = acm_facts.get_certificates(client, module,
```
```suggestion
    certificates = acm_info.get_certificates(client, module,
```

jillr reviewed (commented):
Couple minor nits, otherwise lgtm

s-hertel commented on test/integration/targets/aws_acm/tasks/full_acm_test.yml:
```diff
@@ -0,0 +1,457 @@
+            aws_region: "{{ aws_region }}"
+            aws_access_key: "{{ aws_access_key }}"
+            aws_secret_key: "{{ aws_secret_key }}"
+            security_token: "{{ security_token }}"
```
This should use `no_log`

s-hertel commented on lib/ansible/module_utils/aws/acm.py:
```diff
@@ -0,0 +1,209 @@
+    def get_certificate_with_backoff(self, client, certificate_arn):
+        response = client.get_certificate(CertificateArn=certificate_arn)
+        # strip out response metadata
+        return {'Certificate': response['Certificate'],
```
Both of these keys are not guaranteed to exist.

s-hertel replied:
modifying to use response.get('Certificate') etc seems to fix test issues.

matt-telstra replied:
What issue did you see?

[The Boto Docs](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/acm.html#ACM.Client.get_certificate) don't say they may be omitted.

If the certificate doesn't exist or some other failure happens, boto will throw an exception, so we wouldn't get to the lines where the keys are referenced.

Under what circumstances do these keys not exist?

s-hertel replied:
Yeah, the documentation doesn't. I just ran into it with your integration test suite. I don't know the underlying cause. Perhaps the module needs a waiter if both keys are always expected. I didn't test it to see if it would make a difference. Here's the traceback (Certificate did not exist but CertificateChain did):
```
  File "/tmp/ansible_aws_acm_info_payload_2_c1vcul/ansible_aws_acm_info_payload.zip/ansible/module_utils/aws/acm.py", line 115, in get_certificates
  File "/tmp/ansible_aws_acm_info_payload_2_c1vcul/ansible_aws_acm_info_payload.zip/ansible/module_utils/cloud.py", line 153, in retry_func
  File "/tmp/ansible_aws_acm_info_payload_2_c1vcul/ansible_aws_acm_info_payload.zip/ansible/module_utils/cloud.py", line 140, in retry_func
  File "/tmp/ansible_aws_acm_info_payload_2_c1vcul/ansible_aws_acm_info_payload.zip/ansible/module_utils/aws/acm.py", line 77, in get_certificate_with_backoff
KeyError: 'Certificate'
```

matt-telstra replied:
If I use `.get` instead of `[]`, it will return `None`. That doesn't really fix the problem. It just means we'll run into the problem later, in a way that's harder to debug.

I have added `KeyError` to the `Except` clause in the function which calls this one. Then it throws an error nicely with `module`. Is that acceptable?

jillr replied:
Since we don't know for sure why this happens, and we can't keep testing until we resolve the quota issue, it would be nicer to have a waiter function in case it's just a matter of us being too fast for the API.  That would be in:
https://github.com/ansible/ansible/tree/devel/lib/ansible/module_utils/aws/waiters.py
And you would add something like `CertificateExists` that checks for both the `Certificate` and `CertificateChain` keys to exist.
Then you could use it like https://github.com/ansible/ansible/tree/devel/lib/ansible/modules/cloud/amazon/ec2_vpc_vpn.py#L474 to retry a few times until the keys are available - and if both keys are not returned before the waiter hits `MaxAttempts`, then the `KeyError` exception would handle it.

matt-telstra replied:
Marking as resolved. Line 117 fixes this issue.

matt-telstra committed:
undo copyright addition to wrong file

matt-telstra committed:
fix invalid log message in aws_acm

Co-Authored-By: Jill R <4121322+jillr@users.noreply.github.com>

matt-telstra committed:
rephrase aws_acm_info doc from facts to information

Co-Authored-By: Jill R <4121322+jillr@users.noreply.github.com>

matt-telstra committed:
rename aws_facts var to aws_info

matt-telstra committed:
remove case insensitivity for aws_acm pem compare

matt-telstra commented:
Hmm, it appears that the Shippable tests are all failing now because of a merge conflict.

What's the best way to fix this? I thought the point of GitHub Pull Requests was that GitHub does all the rebasing for me. Do I need to pull `devel`, rebase, and push to my feature branch?

jillr commented on lib/ansible/modules/cloud/amazon/aws_acm.py:
```diff
@@ -0,0 +1,356 @@
+# should result in only errors where equal certs are considered different,
+# in which case the only impact is that changed=True when it should be changed=False.
+# The other kind of error (different certs being considered equal) should not happen.
+def standardize_pem(pem):
```
Sorry I think my comment got buried with @felixfontein's now resolved change request.
I believe you should be able to affect what `standardize_pem` and `pem_compare` are doing using `module_utils.crypto.get_fingerprint()` and comparing fingerprints.  I'd recommend using that over rolling your own pem checker here.

felixfontein replied:
Using `module_utils.crypto.get_fingerprint()` is overkill IMO. It's better to convert the PEM to DER (binary format without the ASCII armour) and then compare; this can already be handled by the Python standard library: https://docs.python.org/3/library/ssl.html#ssl.PEM_cert_to_DER_cert (Should probably also work for Python 2.6, definitely works for 2.7.)

BTW, all code handling PEM files assumes that there is only one PEM object in there, and will use the first one. So using it for a chain won't give the results you want. In this module, this function is also used for comparing chains, so something different is needed anyway.

matt-telstra replied:
The long `failed_when` conditions are used when I have `with_items`. If I split it up into an `assert`, then it gets messy, and then I'll have even deeper `item.item.item.item`s.

How about I split it out into `assert` tasks, *and* use a `vars` block to define `prev_task`, `original_cert`, etc. Will that make things clearer?

jillr reviewed (changes requested):
Giving this a second look and I wonder if the style of the tests is very maintainable. They cover a lot of bases - thanks very much for that.  But the way some vars are named and structured (like `item.item.foo`) can make what's happening difficult to follow, and lengthy `failed_when` blocks (vs a list of assert statements) can make for very dense reading.  While I can follow them, the more I think about it the more I worry about long term maintenance.

The code generally looks good but if possible I'd like to see the tests refactored to be easier to maintain and debug.

s-hertel commented on test/integration/targets/aws_acm/tasks/full_acm_test.yml:
```diff
@@ -0,0 +1,457 @@
+    with_items: "{{ local_certs }}"
+
+  # now upload that certificate
+  - name: upload certificates first time
```
I had to add a short pause before this task to get this task to pass. Perhaps add `retries: ...` or a pause task.

matt-telstra replied:
Like this?

```
  # now upload that certificate
  - name: upload certificates first time
    aws_acm:
      name_tag: "{{ item.name }}"
      <<: *aws_connection_info
      certificate: "{{ lookup('file', item.cert ) }}"
      private_key: "{{ lookup('file', item.priv_key ) }}"
      state: present 
    register: upload
    with_items: "{{ local_certs }}"
    until: upload is succeeded
    retries: 5
    delay: 10
 ```

s-hertel replied:
Yeah, that looks like it would fix it. Thanks!

s-hertel commented on test/integration/targets/aws_acm/tasks/full_acm_test.yml:
```diff
@@ -0,0 +1,457 @@
+  - name: ensure absent cert which doesn't exist - first time
+    aws_acm:
+      <<: *aws_connection_info
+      name_tag: "{{ item.name }}"
```
use the "{{ resource_prefix }}" var as part of all the names. This makes it easy to identify which test created which resource and keeps different test runs from interfering with each other (because the tests are run in parallel in CI).

matt-telstra replied:
Ok. I just added it to the variable list, rather than the `name_tag` lines.

matt-telstra committed:
add no_log for aws_acm credential setting

matt-telstra committed:
add per-test prefix to aws_acm test resource names

matt-telstra commented:
@jillr 

I rebased locally. Now when I try to push I get a clash. Is it safe to use the `-f` flag? (As in, will GitHub retain the same pull request conversation, same change requests etc?)

felixfontein commented on lib/ansible/modules/cloud/amazon/aws_acm.py:
```diff
@@ -0,0 +1,356 @@
+
+
+# For converting PEM bodies to something that can be easily compared
+# e.g. ignore case, strip trailing whitespace
```
You still mention ignore case here.

matt-telstra replied:
I've replaced my custom compare function with a module_utils one.

felixfontein commented on lib/ansible/modules/cloud/amazon/aws_acm.py:
```diff
@@ -0,0 +1,356 @@
+    for c in [' ', '\n', '\t']:
+        pem = pem.replace(c, '')
+
+    # I'm not sure whether the number of dashes at the top and bottom can vary
```
Valid PEM files can have 5 or 6 dashes. Depending on what kind of PEM format you get. Collapsing them to one is definitely a good idea.

matt-telstra replied:
I've replaced my custom compare function with a module_utils one.

felixfontein commented on test/units/modules/cloud/amazon/test_aws_acm.py:
```diff
@@ -0,0 +1,81 @@
+    assert(pem_compare(a, c))
+    assert(pem_compare(b, c))
+
+    assert(not pem_compare(a, d))
```
The certificate can be the same even though the headers are different. The content is important, not the headers. So this should be
```suggestion
    assert(pem_compare(a, d))
```

matt-telstra replied:
I've replaced my custom compare function with a module_utils one.

matt-telstra committed:
make aws_acm use crypto module_util

matt-telstra committed:
clarify copyright for aws_acm

matt-telstra committed:
make aws_acm int test clearer

felixfontein commented on lib/ansible/modules/cloud/amazon/aws_acm.py:
```diff
@@ -52,6 +52,7 @@
     this module will delete all ACM resources in this AWS region with a corresponding C(Name) tag.
     If there are none, it will succeed without effect.
   - Note that this may not work properly with keys of size 4096 bits, due to a limitation of the ACM API.
+  - This module uses the C(cryptography) python library to interact with OpenSSL. (C(pip install cryptography))
```
You need to add it to the module requirements, then.

felixfontein replied:
You also need to make sure that `cryptography` is actually there, and fail gently if it is not. Right now, the module will crash if `cryptography` is not installed when `get_fingerprint_from_pem_cert` is called.

felixfontein replied:
And finally, the most important point, `get_fingerprint_from_pem_cert` won't help you for certificate chains at all! (Except if they have length 1.)

matt-telstra replied:
1. I just added it to the requirements

2. Ah true. I thought the module_util handles that. But upon closer inspection it just carries on. Can you show me an example of a graceful failure? Should I try to `import cryptography` inside the main handler with a `try`?

3. Ugh, yes that's true. I can't see any function in `cryptography` or `crypto.py` which handles chains. Damn. I don't actually know much about the syntax. Last I saw, there's some extra stuff at the start of each cert. Hmm, maybe I really do need to look up the X.509 standard and write my own proper parser.

felixfontein replied:
For 2: see https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/crypto/openssl_privatekey.py#L230-L245 and https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/crypto/openssl_privatekey.py#L683-L685

For 3: that's because there is no such function :) Chains are simply concatenated PEM files. Definitely don't write your own parser, please :) That's never a good idea, especially not for complex beasts such as X.509. (Guess how many security vulnerabilities have been introduced in software when people wrote their own X.509 parser in C/C++ or other memory-unsafe languages...)

The best approach is to look for lines starting with `-----`, and just concatenate the lines between them (between the first and second, between the third and fourth, ...) and concatenate them (with spacing and newlines removed) - so similarly as you were doing in an earlier version of this PR. Or you use code similarly to https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/crypto/acme/acme_certificate.py#L709-L723 to split up the chain into PEM certificates (and then process them with `cryptography` or `ssl.PEM_cert_to_DER_cert` (https://docs.python.org/3/library/ssl.html#ssl.PEM_cert_to_DER_cert, no extra dependency needed).

matt-telstra replied:
Ok, one of the real certificate chains I'm dealing with looks like

```
subject=/C=AU/ST=Victoria/L=Melbourne/O=Telstra Corporation Limited/OU=Telstra Energy/CN=dev.energy.inside.telstra.com
issuer=/C=BM/O=QuoVadis Limited/CN=QuoVadis Global SSL ICA G3
-----BEGIN CERTIFICATE-----
MIIIHTCCB...g6Z5Q=
-----END CERTIFICATE-----

subject=/C=BM/O=QuoVadis Limited/CN=QuoVadis Global SSL ICA G3
issuer=/C=BM/O=QuoVadis Limited/CN=QuoVadis Root CA 2 G3
-----BEGIN CERTIFICATE-----
MIIGF...KmnySglYc
-----END CERTIFICATE-----

subject=/C=BM/O=QuoVadis Limited/CN=QuoVadis Root CA 2 G3
issuer=/C=BM/O=QuoVadis Limited/CN=QuoVadis Root CA 2 G3
-----BEGIN CERTIFICATE-----
MIIFYDCCA...mD3n+M
-----END CERTIFICATE-----

```

Are you saying I should ignore the `subject` and `issues` lines?

matt-telstra replied:
Can I just compare two DER bytes arrays with `=`? There's nothing fancy I need to do? (like `.strip()`)

matt-telstra replied:
I've rewritten my code, using that `ssl` library function to convert each cert to DER. I'm assuming I can just compare DER with `==`.

Please have a look and let me know if my solution is good.

matt-telstra committed:
add explicit crypto dependency to aws_acm

felixfontein commented on lib/ansible/module_utils/crypto.py:
```diff
@@ -132,7 +132,9 @@ def get_fingerprint_of_bytes(source):
 
 def get_fingerprint(path, passphrase=None):
-    """Generate the fingerprint of the public key. """
+    """Generate the fingerprint of the public key. 
```
```suggestion
    """Generate the fingerprint of the public key from the private key."""
```
(And remove the next line.)

matt-telstra replied:
done

goneri commented on test/integration/targets/aws_acm/tasks/main.yml:
```diff
@@ -0,0 +1,33 @@
+      name: virtualenv
+  
+  - pip:
+      name:
```
What dependency is actually in the base image?

matt-telstra replied:
What do you mean by "base image"?

The dependencies I need are a few lines down.

```
        - 'botocore<1.13.0,>=1.12.211'
        - boto3
        - coverage
        - jinja2
        - pyyaml
        - 'pyopenssl>=0.15'
        - 'cryptography>=1.6'
```

goneri replied:
Oh, sorry, I mean "NOT in the base image", all these dependencies should already be available. The although versions may not match.

You can use the `setup_openssl` role to pull OpenSSL, e.g: test/integration/targets/openssl_certificate_info/meta/main.yml

If you can avoid the use of a virtualenv, it would speed up the test execution and it would be more resilient to temporary issue (e.g: pip failure).

matt-telstra replied:
I originally used `setup_openssl`. As per [this thread](https://github.com/ansible/ansible/pull/60552#discussion_r314266062) with @willthames , we decided to switch to the virtual env approach instead, to avoid 

> ERROR: Test target "aws_acm" has changed the test environment!

goneri replied:
Thank you @matt-telstra for the pointer.

goneri commented on test/units/modules/cloud/amazon/test_aws_acm.py:
```diff
@@ -0,0 +1,74 @@
+# (c) 2017 Red Hat Inc.
```
Why 2017?

matt-telstra replied:
Ah, I just copied and pasted another file as a boilerplate for my test. I'll update that.

matt-telstra replied:
Fixed.

matt-telstra commented:
I'm ending up with a lot of commits for this PR. Should I do a squash commit?

felixfontein commented:
If you squash commits, please only squash related ones (like the linting ones). Since the discussion on PEM comparison is not done yet, it is good if the commits changing the code are still around.

matt-telstra committed:
change requests for aws_acm pr

matt-telstra committed:
fix wrong copyright owner aws_acm test

matt-telstra committed:
fix wrong copyright owner aws_acm test

matt-telstra committed:
rewrite aws_acm cert chain compare with regex, no dependency

matt-telstra committed:
fix linting for aws_acm unit test

matt-telstra committed:
fix linting for aws_acm unit test

matt-telstra commented:
Ok, I've rewritten the cert comparison function without any dependencies, using regex and the standard `ssl` library. So now there are no dependencies, and it works with chains.

I've added unit tests for this.

So I think every issue is now resolved?

(Other than waiting for CICD to catch up with the new IAM permissions, which I believe I have added to all the relevant files.)

jillr commented on test/integration/targets/aws_acm/tasks/full_acm_test.yml:
```diff
@@ -101,49 +108,78 @@
+    vars:
+      fetch_after_up_result: "{{ item }}" # corresponding result from task registered as fetch_after_up
+      upload_result: "{{ item.item }}" # corresponding result from task registered as upload
+      original_cert: "{{ item.item.item }}"
```
This really does help a lot with readability, thanks very much @matt-telstra

matt-telstra committed:
fix linting and duplicate ignore

matt-telstra commented:
Why to Ansibot think I'm contributing more than 1 new module?

matt-telstra committed:
fix failed cert chain split in aws_acm, add more tests

matt-telstra committed:
remove errant file

matt-telstra commented:
Ah, I think Ansibot thought I was adding a second new module because of an errant IDE file I added accidentally. I've removed it now.

matt-telstra committed:
more linting fixes for aws_acm

matt-telstra committed:
fix sanity ignore

felixfontein commented on lib/ansible/module_utils/crypto.py:
```diff
@@ -152,6 +152,16 @@ def get_fingerprint(path, passphrase=None):
     return get_fingerprint_of_bytes(publickey)
 
 
+def get_fingerprint_from_pem_cert(data):
```
Since you're no longer using this function, please remove it.

matt-telstra replied:
done

felixfontein commented on lib/ansible/modules/cloud/amazon/aws_acm.py:
```diff
@@ -0,0 +1,396 @@
+from ansible.module_utils.aws.core import AnsibleAWSModule
+from ansible.module_utils.ec2 import boto3_conn, ec2_argument_spec, get_aws_connection_info
+from ansible.module_utils.aws.acm import ACMServiceManager
+from ansible.module_utils.crypto import get_fingerprint_from_pem_cert as get_fingerprint
```
Also remove this import, you don't use it.

matt-telstra replied:
done

felixfontein commented on lib/ansible/modules/cloud/amazon/aws_acm.py:
```diff
@@ -0,0 +1,396 @@
+    # Use regex to split up a chain or single cert into an array of base64 encoded data
+    # Using "-----BEGIN CERTIFICATE-----" and "----END CERTIFICATE----"
+    # Noting that some chains have non-pem data in between each cert
+    expr = re.compile(r"-+BEGIN\s+CERTIFICATE-+([a-zA-Z0-9\+\/=\s]+)-+END\s+CERTIFICATE-+")
```
You don't need that regex here.

matt-telstra replied:
done

felixfontein commented on lib/ansible/modules/cloud/amazon/aws_acm.py:
```diff
@@ -0,0 +1,396 @@
+# Noting that some chains have non-pem data in between each cert
+# This function returns only what's between the headers, excluding the headers
+def pem_chain_split(pem):
+    expr = re.compile(r"-+BEGIN\s+CERTIFICATE-+([a-zA-Z0-9\+\/=\s]+?)-+END\s+CERTIFICATE-+")
```
Better store that in a global variable, otherwise there's no need to compile it before using it.

felixfontein replied:
About the expression: I would change it to
```suggestion
    expr = re.compile(r"------?BEGIN [A-Z0-9. ]*CERTIFICATE------?([a-zA-Z0-9\+\/=\s]+)------?END [A-Z0-9. ]*CERTIFICATE------?")
```
That covers some more corner cases (https://tools.ietf.org/html/rfc7468#appendix-A).

(Also see https://stackoverflow.com/a/5356351)

matt-telstra replied:
done

felixfontein commented on lib/ansible/modules/cloud/amazon/aws_acm.py:
```diff
@@ -0,0 +1,396 @@
+    prefix = "-----BEGIN CERTIFICATE-----"
+    suffix = "-----END CERTIFICATE-----"
+    pem_full = '\n'.join([prefix, pem, suffix])
+    return(PEM_cert_to_DER_cert(pem_full))
```
You can also simply use `base64.b64decode()` instead of first re-adding the headers and then calling `PEM_cert_to_DER_cert()`. Also, you should check whether `PEM_cert_to_DER_cert()` (or `base64.b64decode()` if you use that) raises an exception, and react to it if necessary.

(BTW, this is how `PEM_cert_to_DER_cert()` can be implemented: https://github.com/python/cpython/blob/master/Lib/ssl.py#L1456-L1467)

felixfontein replied:
With Python 3.6, I've seen `ValueError` (incorrect header) and `binascii.Error` (bad Base64).

matt-telstra replied:
I just changed it to `b64decode`.

matt-telstra committed:
rewrite cert compare in aws_acm to use base64 decode

matt-telstra committed:
improve regex for pem cert chain split in aws_acm

matt-telstra committed:
undo changes to crypto module util for aws_acm

felixfontein commented on lib/ansible/modules/cloud/amazon/aws_acm.py:
```diff
@@ -0,0 +1,403 @@
+    this module will delete all ACM resources in this AWS region with a corresponding C(Name) tag.
+    If there are none, it will succeed without effect.
+  - Note that this may not work properly with keys of size 4096 bits, due to a limitation of the ACM API.
+version_added: "2.9"
```
This needs to be updated to `2.10`, otherwise CI will complain on the next run.

matt-telstra replied:
changed

felixfontein commented on lib/ansible/modules/cloud/amazon/aws_acm.py:
```diff
@@ -0,0 +1,403 @@
+    chain_b_pem = pem_chain_split(module, b)
+
+    if len(chain_a_pem) != len(chain_b_pem):
+        return(False)
```
Personally, I don't think this is good Python style. `return` is not a function, it's a keyword.
```suggestion
        return False
```

matt-telstra replied:
Ok, I just changed it.

felixfontein reviewed (approved):
Just commenting on chain comparison. I think it looks good now.

Personally, I would do error handling on a higher level and not necessarily fail the module in the code, but for this module this is ok. If this is moved to `module_utils`, it should definitely be changed.

matt-telstra committed:
increment ansible version for new aws_acm module

matt-telstra committed:
convert aws_acm return(x) to return x

matt-telstra committed:
increment version added for aws_acm_info new features

matt-telstra committed:
fix linting

matt-telstra commented:
How long do I need to wait until the CICD system catches up to the new permissions required for integration testing? Which human is going to be looking at my PR, and applying the updated IAM policy?

matt-telstra commented:
Sorry, that may have come off as rude. I'm not trying to be rude. Just curious.

s-hertel commented:
@matt-telstra I'd like another set of eyes on the CI change before that gets merged because I'm still a little confused over https://github.com/mattclay/aws-terminator/pull/49#discussion_r318362340. @jillr Do you mind taking a look at https://github.com/mattclay/aws-terminator/pull/49 and testing it? And do you mind updating your review on this one as well?

jillr commented on lib/ansible/modules/cloud/amazon/aws_acm.py:
```diff
@@ -0,0 +1,403 @@
+
+            # Are the existing certificate in ACM and the local certificate the same?
+            same = True
+            same &= chain_compare(old_cert['certificate'], module.params['certificate'])
```
```suggestion
            same &= chain_compare(module, old_cert['certificate'], module.params['certificate'])
```
And the 2 below uses.

mattclay commented on test/sanity/ignore.txt:
```diff
@@ -1135,19 +1131,19 @@ lib/ansible/modules/cloud/azure/azure_rm_devtestlabvirtualmachine.py validate-mo
 lib/ansible/modules/cloud/azure/azure_rm_dnsrecordset.py validate-modules:parameter-type-not-in-doc
 lib/ansible/modules/cloud/azure/azure_rm_dnsrecordset.py validate-modules:doc-missing-type
-lib/ansible/modules/cloud/azure/azure_rm_dnsrecordset_facts.py validate-modules:parameter-type-not-in-doc
+lib/ansible/modules/cloud/azure/azure_rm_dnsrecordset_info.py validate-modules:parameter-type-not-in-doc
```
These changes are unrelated to the PR. Perhaps there was an issue rebasing?

matt-telstra replied:
Yeah probably. I'll fix that now.

matt-telstra replied:
Fixed

jillr reviewed (commented):
Apologies for the delays @matt-telstra, the Certificate quota we've hit on our CI account (and our personal AWS accounts now too...) is making it really challenging to progress this PR but we will get that sorted with AWS!  This patch is looking really good.

matt-telstra commented:
What is the exact quote limit you're hitting?

Is it total number of certs, or the *rate* at which you can import certs?

jillr commented:
> What is the exact quote limit you're hitting?
> 
> Is it total number of certs, or the _rate_ at which you can import certs?

That's the strange part - it doesn't seem to be a documented limit but we've hit it on 3 different AWS accounts.  It reports as total uploaded certs.

```
TASK [aws_acm : upload chained cert, different chain] ***************************************************************************************************************************
task path: /home/jill/.ansible/test/tmp/aws_acm-flyQQq-ÅÑŚÌβŁÈ/test/integration/targets/aws_acm/tasks/full_acm_test.yml:435
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: LimitExceededException: An error occurred (LimitExceededException) when calling the ImportCertificate operation: You have imported the maximum number of 20 certificates in the last year.
fatal: [testhost]: FAILED! => {"boto3_version": "1.9.223", "botocore_version": "1.12.223", "changed": false, "error": {"code": "LimitExceededException", "message": "You have imported the maximum number of 20 certificates in the last year."}, "msg": "Couldn't upload new certificate: An error occurred (LimitExceededException) when calling the ImportCertificate operation: You have imported the maximum number of 20 certificates in the last year.", "resource_actions": ["acm:DescribeCertificate", "acm:ListCertificates", "acm:ImportCertificate", "acm:ListTagsForCertificate", "acm:GetCertificate"], "response_metadata": {"http_headers": {"connection": "close", "content-length": "121", "content-type": "application/x-amz-json-1.1", "date": "Thu, 05 Sep 2019 22:10:36 GMT", "x-amzn-requestid": "f36563b7-3a8a-48ed-849c-0f4dfd2fcf39"}, "http_status_code": 400, "request_id": "f36563b7-3a8a-48ed-849c-0f4dfd2fcf39", "retry_attempts": 0}}
```
In none of our accounts can we find evidence of a certificate or ACM related quota with a value of 20, any AWS documentation of such (the ACM docs say we should get 1k certs, with 2k uploads per year), or a quota that shows exhausted utilization.  So we're going off some forum posts that other folks who have run into this same thing have chased it with AWS support and unspecified solutions on the AWS side enabled further use of ACM.  @mattclay has access to work with AWS support but unfortunately with 2.9 releasing and AnsibleFest and contributor summit next week everyone is stretched fairly thin right now.

matt-telstra commented:
Any update on the AWS ACM limit?

jillr commented:
Hi @matt-telstra, I'm really sorry but we haven't gotten that sorted out with AWS yet.  What we can do for this PR is take it with the tests listed as unsupported in `test/integration/targets/aws_acm/aliases`, and we have an internal issue already to track the quota issue so we can get them enabled once that's figured out.  The only thing we need still then is for `module` to be passed to `chain_compare` in lib/ansible/modules/cloud/amazon/aws_acm.py at lines 356, 360, and 364.  Thanks so much for your patience and our apologies for the difficulties here.

matt-telstra committed:
fix bugs with AWS ACM

matt-telstra commented:
I have just rebased to the `devel` branch. It seems there were a few changes to `aws_acm_info`, but I think the merge went fine.

I've added `module` as an argument to `chain_compare`.

I just re-ran my integration tests, and saw that error someone saw earlier, where boto just doesn't return `Certificate` as an attribute. The cause was that in some states (e.g. `VALIDATION_TIMED_OUT`) there is no actual certificate. So `get_certificate` returns no certificate. I've added an if statement to check for that. Now tests passed. (I couldn't reproduce the problem before because the other certificates I happened to have in my account didn't have that status.)



Where do I specify that a test is unsupported and can be ignored?

Let me know if I should squash the commits yet or not.

felixfontein commented:
Your rebase didn't work well with the `test/sanity/ignore.txt` file. You added **a lot** of lines there.

> Where do I specify that a test is unsupported and can be ignored?

Simply add a new line with `unsupported` in it to `the `aliases` file.

felixfontein commented:
Also there's a failing sanity test: https://app.shippable.com/github/ansible/ansible/runs/149151/4/tests

matt-telstra committed:
fix bad rebase

matt-telstra committed:
disable AWS ACM integration test, due to AWS account limit issue

matt-telstra commented:
Ok, I *think* I fixed the rebase. Now `ignore.txt` is not appearing in the Github summary.

matt-telstra commented:
I've added an 'unsupported' line too. Now I'll just wait to see if the linting checks all pass.

matt-telstra commented:
> test/integration/targets/aws_acm/aliases:0:0: conflicting alias `shippable/aws/group[1-2]` and `unsupported`

Ok, what does this mean?

Do I need to remove the `shippable` line from the Alias file?

```
cloud/aws
shippable/aws/group2
aws_acm_info
unsupported
```

matt-telstra committed:
remove aws acm integration test from shippable group

mattclay commented:
I've opened a support ticket to find out if there's anything we can do about the limit.

matt-telstra commented:
Ok the tests have all passed. @jillr it says that you've made one change request. I can't find it in this long thread. Is there anything I haven't addressed yet?

@mattclay I have also lodged a support ticket. I'm on the premium support plan. It's normally extremely helpful and prompt, but this particular issue has taken them months and they keep sending me in circles.

jillr reviewed (approved):
@matt-telstra You made the requested changes from that review (the missing `module` args), github just requires me to manually dismiss or update my review. Everything looks good here, and I was able to get one last test in with everything passing. Thanks again for all your work on this one.

matt-telstra commented:
Woo hoo!

mattclay commented:
AWS support has corrected the limit on our account. Our limit should now be 2,200 certificate imports per year. I don't know what our burn rate is, so I don't know if that will be adequate or not. I'll see if support is able to tell us how many we've currently used.

matt-telstra commented:
Here's what AWS support just said to me about my account:

> Currently, we cannot disclose any firm limits for requests per second on the ImportCertificate API call; the limits you are encountering are in place to protect the infrastructure of the service from issues, and as a result we are unable to increase this request rate any further. We do still recommend applying exponential backoff and retry mechanisms to any workflow that is performing ImporCertificate but getting throttled. We will consider adding these API limits to the documentation if possible at a later date.

> Regarding the number of imported certificates per year, in the past 90 days (the extent of what I can look back on), there have been 338 successful ImportCertificate API calls in ap-southeast-2, and there are currently 55 imported certificates in the region. It's possible that you may be hitting the Imported Certificates Per Year limit if this has been how you've been importing certificates regularly throughout this year, but I can't confirm if that's the trend or not. The ACM team has upped your limit from 2000 to 2500 in order to unblock you, but they also recommend rethinking how you import your certificates into AWS. For example, **if you import into the same certificate ARN as another cert, the newly imported certificate will not count towards your maximum for the year.**

> If you have any other questions or issues, please let us know. If you'd like to discuss how to better architect your workflow for how you use certs, you can engage us or your TAM if you are interested in talking to a Solutions Architect about how to optimize your environment with respect to ACM. Thank you for your time and patience as we looked into your issue, and we're here if you need anything else!

Always using the same ARN would be good, but I don't know how we can do that *and* fix [this issue](https://github.com/ansible/ansible/pull/64537#pullrequestreview-313047978).
